### ISSUE TITLE
System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning() silently swallows error conditions

### ISSUE THREAD

**Issue by Dmitry-Me** [dotnet/corefx#6119]
`System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning()` has a Debug.Assert check that the errors connection is not empty and also this excellent `if`:

     SqlConnection exception = null;
     //later...
     if (temp != null && temp.Count > 0)
          //craft exception
     }

     //later ...
     if (exception != null ) {
         //do a lot of cool stuff
     } // else do nothing

which means that allowing to have this function called when there're no elements in `temp` leads to a logical error - no error indication is given and the connection is just silently closed.

Those `Debug.Assert`s should instead throw something like `InvalidOperationException`.

**Comment by saurabh500** [comment]
@Dmitry-Me What is the issue here? Are you asking to change what the Debug assert failure does  because there is some chance that this function can be incorrectly invoked from a location where there may be nothing in `temp` ?

**Comment by Dmitry-Me** [comment]
@saurabh500 Yes, the current implementation allows for error hiding. With assertions disabed it can called in scenarios where it silently does nothing.

**Comment by JonHanna** [comment]
Can you show a repro where that happens?

**Comment by Dmitry-Me** [comment]
@JonHanna No, I cannot. This is "what if the Moon in in wrong phase" kind of defect.
